core/arxiv/submission/domain/event/__init__.py

```diff
@@ -0,0 +1,1263 @@
+
+    def validate(self, submission: Submission) -> None:
+        """Cannot apply to a finalized submission."""
+        validators.submission_is_not_finalized(self, submission)
+
+    def project(self, submission: Submission) -> Submission:
+        """Update :attr:`.Submission.submitter_contact_verified`."""
+        submission.submitter_contact_verified = True
+        return submission
+
+
+@dataclass()
+class ConfirmAuthorship(Event):
+    """The submitting user asserts whether they are an author of the paper."""
+
+    NAME = "confirm that submitter is an author"
+    NAMED = "submitter authorship status confirmed"
+
+    submitter_is_author: bool = True
+
+    def validate(self, submission: Submission) -> None:
+        """Cannot apply to a finalized submission."""
+        validators.submission_is_not_finalized(self, submission)
+
+    def project(self, submission: Submission) -> Submission:
+        """Update the authorship flag on the submission."""
+        submission.submitter_is_author = self.submitter_is_author
+        return submission
+
+
+@dataclass(init=False)
+class ConfirmPolicy(Event):
+    """The submitting user accepts the arXiv submission policy."""
+
+    NAME = "confirm policy acceptance"
+    NAMED = "policy acceptance confirmed"
+
+    def validate(self, submission: Submission) -> None:
+        """Cannot apply to a finalized submission."""
+        validators.submission_is_not_finalized(self, submission)
+
+    def project(self, submission: Submission) -> Submission:
+        """Set the policy flag on the submission."""
+        submission.submitter_accepts_policy = True
+        return submission
+
+
+@dataclass()
+class SetPrimaryClassification(Event):
+    """Update the primary classification of a submission."""
+
+    NAME = "set primary classification"
+    NAMED = "primary classification set"
+
+    category: Optional[taxonomy.Category] = None
+
+    def validate(self, submission: Submission) -> None:
+        """Validate the primary classification category."""
+        validators.must_be_a_valid_category(self, self.category, submission)
+        self._creator_must_be_endorsed(submission)
+        self._must_be_unannounced(submission)
+        validators.submission_is_not_finalized(self, submission)
+        validators.cannot_be_secondary(self, self.category, submission)
+
+    def _must_be_unannounced(self, submission: Submission) -> None:
+        """Can only be set on the first version before publication."""
+        if submission.arxiv_id is not None or submission.version > 1:
+            raise InvalidEvent(self, "Can only be set on the first version,"
+                                     " before publication.")
+
+    def _creator_must_be_endorsed(self, submission: Submission) -> None:
+        """Creator of this event must be endorsed for the category."""
+        if isinstance(self.creator, System):
+            return
+        try:
+            archive, _ = self.category.split('.', 1)
```

Comment:
Consider using `arxiv.taxonomy.CATEGORIES` to determine the archive from the category, even though this should work (but may not always).

Reply:
fixed in f3c02d6